### ISSUE TITLE
[Question] Prettier Philosophy and Rationale printWidth

### ISSUE THREAD

**Issue by jogelin** [prettier/prettier#4298]
I am opening an issue just to get an information about one decision, not related to a specific code formatting rule, but about a core concept in prettier : the **`printWidth`** option.

First of all, I really love prettier and the philosophy ! I configured it in all projects in my company and I am using it every day ! 

But....why, in 2018, are we talking about **`printWidth`**  when we talk about formatting ? Are you still printing your code ? Are you all using the same screen resolution/scale ?

If it's about removing horizontal scroll, the best option is to use a “soft wrap” setting/plugin in your editor...it will break your line visually and it will adapt this automatically if you reduce the width of your editor.

The philosophy of prettier is to define a common way to format your code and stop debating... Why this common way is impacted by a printWidth ? 

For example, sometimes, my JS array is formatted in one line and sometimes in multiple lines just because of the printWidth….this is not a common way to format arrays…It should be even more strict than that: 
* put my array in one line **all the times** and I'll soft wrapping if it's too long
* or...split my array in multiple lines **all the times**
* or...format my array in one line if it contains one element and split my array in multiple lines if it contain more than one element **all the times**

Maybe I completly misunderstanding the philosophy of prettier but I feel that the **`printWidth`** is not matching it.

What do you think ?

**Comment by jogelin** [comment]
@vjeux I agree that the `printWidth` is a quick solution for today.

My only advice would be to change our mindset in formatting. Prettier already started this changement by pushing people to use one way of working...but I think the strategy should go even deeper in this idea by removing the printWith and more focusing on syntax.

Developers are using different screens and different resolutions but they still need a common way to format their codes.

Indeed, I don't have the truth, I just want to make you thinking about this. I like prettier and I think it's important that it going on a good direction ;)

**Comment by vjeux** [comment]
Oh, trust me I’ve been thinking about this for a very long time. The fact that as an industry we’re still using 80-100 columns tells me that this is not just because of some old terminals but because it has some good inherent properties. Maybe it’s not the perfect way to print code but so far it has worked well and is how in practice people format most of their code. The fact that prettier has become so popular so quickly shows that the current system formats code that looks like how people format their xode in practice.

**Comment by kachkaev** [comment]
+1 to git diff argument. The longer the line, the bigger the diff and the harder it is to track what has been chanced. Soft wrap won't help here.

**Comment by jogelin** [comment]
@lydell @vjeux @kachkaev My point is not to say "The code should not be split", my point is "To keep consistency, the code should be split based on syntax, not on print width". I think you will meet a lot of issue like #4172

**Comment by atombender** [comment]
I'm not entirely sure if this is the point that @jogelin wants to get across, and if this is the right issue, but I'm consistently disappointed by the heuristic that Prettier uses to determine whether to wrap something, especially in JSX blocks.

Here is an example (1) of an absurdly unnecessarily "vertical" orientation:

```javascript
// Prettier
{item.mainImage && (
  <img
    src={buildImageUrl
      .image(item.mainImage)
      .height(400)
      .url()}
  />
)}

// How I expected it
{
  item.mainImage &&
    <img src={buildImageUrl.image(item.mainImage).height(400).url()}/>
}

```

Whereas this (example 2) is somehow considered readable:

```javascript
// Prettier
{item.year != null && <div className="Item-about-title-year">{item.year}</div>}

// How I wanted it
{
  item.year != null &&
    <div className="Item-about-title-year">
      {item.year}
    </div>
}
```

I haven't read the code, it seems to me that Prettier uses a single condition — whether a line will break the "print width" — to determine whether to optimize for a horizontal or vertical layout, and doesn't actually have a readability-oriented heuristic layout algorithm at all.

In example 1, the wrapping severely impedes readability, and I can imagine a heuristic scorer that could rate it low because (1) the tag that is broken off is short (so why wrap the second line?), (2) it splits a JSX attribute expression (usually a bad idea since it causes attributes to not align underneath each other), and (3) individually, the split lines are very short, creating a "clustered" look.

In example 2, the wrapping also impedes readability, and a scorer should score it love because (1) it's a conditional, so the form `{condition && tag}` is almost always better written on two separate lines, (2) the div is a block container, so it has more "weight" than something inline like a span, and (3) since it's a block container, its contents should also be separated, unlike something inline.

I see issues like #4210 and it's obvious that a heuristic-based formatter would work better. "Array of many small things" is a candidate for a "matrix" renderer, whereas "array of many complicated things" is a candidate for a one-per-line renderer.

**Comment by vjeux** [comment]
Thanks for the concrete examples

> it's obvious that a heuristic-based formatter would work better

There are a ton of heuristics inside of prettier already, the issue is what heuristic to actually implement.

For the list of many many ints, we could use the “fill” algorithm. I’m still unclear whether it would actually solve the problem. Usually lists of numbers like this have a “width” that’s not the line length and I’m not sure how to infer that width.

> a scorer should

In my experience, it’s very hard to build this efficiently. If you want to score multiple options this means that you’ve got to compute them. ASTs tend to be deeply nested structures and if you have multiple choices per node, you end up in an exponential complexity. As an implementer of those pretty printers, you end up spending most of your time trying to find optimizations such that a seemingly simple program doesn’t take hours to pretty print.

The other issue is around predictability, you would have in your codebase the same construct being printed in different ways based on context. Prettier does some of that but stays minimal on that front.

Again, I would highly recommend you to try out different approaches and see how it would solve some of the issues we have with prettier.

**Comment by vjeux** [comment]
For the first example, this is actually a place where prettier doesn’t follow 80 columns and instead automatically breaks if there are more than 3 members in a chain. I wasn’t happy to put this in place but so many people complained about it that we needed to do something about it. Since then it’s still an issue but much less so. I agree that the indent looks off, we could put the expression on its own line.

For the second example, prettier would print it closer to the way you want if you used 80 columns: https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgBo7AA6UOOAZgBQCUhJZZATnDAK6OkA8AJgJYBuAPmC94AWwB0ATzjpGOAIQBeHFFapUOAGRacPATjAZsAOXRi4SoiACS4gLToARhFYx7MUajj2Zc68KicJJ+jAC+nAD0fEJhJFExgvQMZGFhIAA0IBAADp7QmMigcowQAO4ACnIIhSjo-BC83JkgTozoYADWLADKOR28UADmyDCMrHBZg5hwjDAV7UNi6Mjk6KgzWQBWmAAeAELtXb3mcAAyg3Cr65sg-YwzjMitzlKo0C05jIMwAOpNMAAFsgABwABiyXwgM1+7Ryzy+cEe-CuWWYAEdWLxmAt0EsVkg1htJiAZmJeKNxiTMIMht4AIqsCDwa7ErIwZz-bhA5AAJnZ7V4qFpAGEIGJls8oNBUSBWDMACrOWpEmZpIA

As for the content inside being split out, we used to split it out but people complained that it made inline blocks like this take up too much space so now we don’t anymore.

So, for those examples, the problem isn’t actually the 80 column logic but heuristics that we did that went against your taste. Not everyone formats code the same way and we try to build something that would look reasonable most of the time.

One thing though is that both of those examples look like a human could have written them, so this passes the “prettier turing test”. It’s actually a really good place to be in where the issues that people report about prettier are no longer about crazy looking code but more nuanced ways to format a piece of code, it means that the project succeeded :)

**Comment by atombender** [comment]
Not sure what you mean by "Turing test". Anything can be written by a human; when Prettier does what I don't want, then it doesn't matter if it passes any such test to another observer.

Prettier is opinionated in a way that is opaque, and that's a problem. The most successful automatic formatter in the history of computing is, arguably, Go's gofmt, and nobody really talks about gofmt because it's opinionated in just the "right amount". Gofmt is opinionated, but Go's most brilliant design decision here was to mostly honour the programmer's own line breaks within declaration blocks. This means that Go will happily accept both of these, and not reformat them:

```go
func foo(a int, b string, c []string) {
  ...
}

func foo(
  a int,
  b string,
  c []string) {
  ...
}
```

This is awesome, because it turns out that forcibly reformatting this to be horizontal or vertical is *not useful*, and it's also a case where it's *nearly impossible for a computer to decide what is right*. Unfortunately, Prettier would choose one of the other here, and it would seem entirely arbitrary, because it's not smart enough to decide what "looks best", and it can't. Adding or removing something could completely flip the algorithm used, and no matter the underlying heuristic, it would be surprising.

Gofmt will, like Prettier, normalize whitespace and alignment and so on, but it will _generally_ leave expression wrapping alone. In my opinion, Gofmt's design is much *wiser* and maps better to a human workflow, because while it will mercilessly prettify tangled code into a canonical form, it also leaves just the right part alone.

Play with Gofmt [in this playground](https://play.golang.org/p/x21ZdRp-NYd), for example. Even if you don't know Go, it should make sense; change the wrapping and click "Format" and see how it adapts.

It turns out that if you go with Gofmt's gentle heuristics, almost all of the concerns about wrapping and orientation melt away. For example, the questions about array packing and so on just disappear, but you choose the style yourself; there's no need for the formatter to try to have an opinion. I say almost because unlike JavaScript, Go typically has much less function chaining (and have different line termination rules), and so the resulting formatting [isn't optimized for that style](https://play.golang.org/p/r2UCawd1TAV). But even so, the problem surface becomes smaller.

**Comment by vjeux** [comment]
Prettier and gofmt made different decisions for this particular part of the syntax. My goal with prettier was to have people stop spending time formatting their code, so the more we can canonicalize, the better. We made some trade-offs where we respect newlines inside of {} and respect at most one empty line in between statements.

I believe that overall it worked well, there are some edge cases like long lists of data and structured shapes like matrices where the output is suboptimal but that’s those are some rare edge cases relative to the total amount of code.

There is no right or wrong solution here, it’s all about trade-offs. I want format on save to do 99% of the job. If you start respecting more of the original formatting this means that you have to let the user do more formatting work.

Gofmt was surely a huge success and a big inspiration for prettier. Now, I wouldn’t be surprised if there were more lines of code formatted with prettier than gofmt ;)

**Comment by atombender** [comment]
It goes beyond mere edge cases, of course. Here's another example. Original code:

```javascript
app.post('/api/v1/things/:id', (req, res) => {
  ...
})
app.post('/api/v1/things/:id/:format', (req, res) => {
  ...
});
```

New code:

```javascript
app.post('/api/v1/things/:id', asyncHandler(async (req, res) => {
  ...
}))
app.post('/api/v1/things/:id/:format', asyncHandler(async (req, res) => {
  ...
}));
```

Looks great. Hit Save and Prettier (with print width 100) gives us:

```javascript
app.post(
  '/api/v1/things/:id',
  asyncHandler(async (req, res) => {
    ...
  })
);
app.post(
  '/api/v1/things/:id/:format',
  asyncHandler(async (req, res) => {
  ...
  })
);
```

Prettier's strategy is, in my opinion, fundamentally developer-unfriendly. The code was fine. Now it looks like a junior intern, or worse, a very stupid robot, went over it. FWIW, I'm constantly running into these over-eager wrapping cases where the wrapping cannot be fixed because Prettier thinks it knows better than humans. Fighting tools isn't fun. Code isn't _really_ a data structure. Prettier would need a scoring heuristic SAT solver or some serious AI to do as well as a human, and that's the Achilles' heel.

> There is no right or wrong solution here, it’s all about trade-offs.

FWIW, Gofmt doesn't have these trade-offs, so objectively Prettier is less "right".

**Comment by vjeux** [comment]
The new code is over 80 columns so if you had a linter and respected it, you would have changed it. If you did so, the way prettier printed it would have been a valid choice.

If you don’t respect the 80 columns lint for this case, then you are likely not the target audience for prettier.

**Comment by vjeux** [comment]
I think that there’s a fundamental mental step you need to do when you use prettier: you don’t write code formatted and then apply prettier, instead you write code in one line with no respect for formatting, press save and then prettier formats it.

If you use the first mindset, then as you mention you are going to fight with the tool which is going to rewrite your carefully crafted code in a different way than you expected. This is infuriating.

But with the second one, prettier is actually a help. You don’t need to think about formatting anymore, there’s a tool that takea this part of the job and automates it.

In order for it to work, the output needs to be good enough in most cases. But doesn’t have to be perfect. If it doesn’t get it exactly right 1% of the time but made you save 20% of coding time, it’s likely worth it.

**Comment by atombender** [comment]
My line width, including `printWidth`, is set to 100, not 80 (because my editor is not a 1970s terminal), so I'm not sure what you're referring to here.

The old code was of course more than 80 wide, too. I assumed it'd be obvious that I'd elided the function body. Just to be clear:

```javascript
// Old code
app.post('/api/v1/things/:id', (req, res) => {
  console.log('this is the function body');
  console.log('this is the function body');
  console.log('this is the function body');
  console.log('this is the function body');
  console.log('this is the function body');
});
// New code, after Prettier
app.post(
  '/api/v1/things/:id',
  asyncHandler((req, res) => {
    console.log('this is the function body');
    console.log('this is the function body');
    console.log('this is the function body');
    console.log('this is the function body');
    console.log('this is the function body');
  })
);
```

The error is that Prettier thinks that a longer argument should cause the _preceding_ lines to be wrapped, which is wrong since they're not the ones violating the line width. I expected the general case to wrap like this:

```
functionCall(arg1, arg2, ...,
  asyncHandler((req, res) => {
    // body
  }));
```

Here's a pathological test case that generalizes the problem:

```javascript
test('01', '02', '03', '04', '05', '06', '07', '08', '08', '09', '10', '11', '12', () => {
  // body
});
test('01', '02', '03', '04', '05', '06', '07', '08', '08', '09', '10', '11', '12', '13', () => {
  // body
});
test(
  '01',
  '02',
  '03',
  '04',
  '05',
  '06',
  '07',
  '08',
  '08',
  '09',
  '10',
  '11',
  '12',
  '13',
  '14',
  () => {
    // body
  }
);
```

What I expected here:

```javascript
test('01', '02', '03', '04', '05', '06', '07', '08', '08', '09', '10', '11', '12', '13',
  '14', () => {
  // body
});
```

Again, Prettier degenerates in ways that don't match human preferences, going from a horizontall fill to a vertical orientation.

> You don’t need to think about formatting anymore ...  If it doesn’t get it exactly right 1%

Until it breaks, which is considerably more than 1% for me. For me, Prettier constantly violates the principle of least surprise, because there is no "one true way" to format code. Again, this is what Gofmt got right and Prettier gets wrong.

**Comment by vjeux** [comment]
I believe that the issue here is that you have a different guidelines on how to break lines that are too big. I don’t know about all the ones that are in use in the JS ecosystem but at least at Facebook, if it doesn’t fit a single line, all the arguments have to be put on their own line (except for the rare edge cases I mentioned). Prettier implements this strategy.

Given prettier extremely fast adoption and the fact that this hasn’t been a hugely debated issue, I think that either it’s the way that a lot of people chose to write code or that it is not so bad that it’s a blocker for using prettier.

One of the advantage that go has is that almost nobody wrote go before it had a formatter, so whatever the formatter did was seen as the canonical way to write go. It was not the case with JavaScript, so prettier is bound to do choices that some people (that can have very sizable codebases and mindshare) did differently. We tried to listen to the biggest ones and add options for them.

I’m sorry that prettier doesn’t format code the way you want. I’m not convinced that doing the same as go would align well with prettier’s vision, so it’s unlikely that we are going to do so.

**Comment by atombender** [comment]
It's a pretty big jump from a horizontally filled argument list to one splattered over 14 lines. That's a weird rule, and not one I've ever seen outside of Prettier. None of the projects whose codebases I have skimmed through recently implement this "Facebook style" (e.g. Vue, Next.js, Express; all important JavaScript projects that predate Prettier). A hanging indentation style, where only overflowing terms are wrapped to the next line unless reorienting things into a vertical list or "matrix" improves readability, has been the norm since IE 4. I mean, look at jQuery, the most ancient of all JavaScript libraries still in use; if you wanted a "standard" style, there it is. Anyway.

Related to that, here's another pathological case that hit me just now (slightly tweaked for brevity):

```
// Original code
const completionHandler = callback || handler(request, {
  error: err => r.report(err, request),
  config: getConfig(),
});

// Prettier
const completionHandler =
  callback ||
  handler(request, {
    error: err => r.report(err, request),
    config: getConfig(),
  });
```

This overly vertical orientation is not more readable than the original code, and only serves to obscure intent; the `||` expression splattered across so many lines when just one will do.

_Of course_ Gofmt had more freedom to set guidelines because it was a greenfield project, but that doesn't in any way refute the arguments that (1) Prettier makes objectively bad choices (more than can be characterized as "edge cases", in my opinion), and (2) these would have been solved by using a similar strategy to Gofmt's instead of Prettier's current brute-force "just ram everything through the canonicalizer" strategy.

**Comment by vjeux** [comment]
For your last example, it should ideally be the way you wrote it. We already do optimizations if you write x || { ... }. In this case the issue is that the {} is inside of a call so it’s harder to make it work within prettier implementation. We may be able to do it.

It’s true that the current set of rules implemented by prettier tends to make things more expanded than not.

This is indeed going against codebases like jQuery that tend to (ab)use a lot of JavaScript patterns to write very terse code. My understanding is that it’s because minifiers at the time were very primitive and they had to use such a coding style to have a final reasonably small output.

Thanks for this discussion, you raised some good points.

One thing I’ve wanted was to make it easy to fork prettier and be able to play with different variants without controlling the core library. If you use a git version of a fork then “hopefully” (I haven’t actually tested it) all the integrations should read that version and you can use it on your project.

**Comment by merlinnot** [comment]
Hey, just wanted to share my opinion on what Prettier currently does by throwing a few emojis at ya:

❤️👌😍🤩👍😎🧠🧚‍♀️💃🌈🥇🏆🏅🚀💡💎🎉💯✅

There is an audience that absolutely loves 80 columns and couldn’t live with 81. Great job!

**Comment by jogelin** [comment]
@atombender I really like your approach !

@vjeux If the philosophy of prettier is only targeting the audience using 80 column (or printWidth). Then I understand your philosophy and I respect it... 

I just wanted to share my point of view about the philosophy's direction of prettier. During my experiences, I meet less and less people that use printWidth in their IDE (@troll yes sometimes, I meet Java developers still working on eclipse with printWidth :p :p), but usually, they used softwrap because of the reasons that I already explained. Usally, I am using git in command line....but not for `git diff`....I am using a UI which is, to me, more clean....

I think this is the way that the development world is going, be more flexible, using tools. 
Again, I really like the idea of having the same way of formatting ! Thx for that !

@lydell I will provide some examples of formatting + rules with the Angular syntax in separate tickets

**Comment by felixfbecker** [comment]
In a perfect world, we would just commit code without whitespace and the editor would use prettier to pretty-print it to the perfect print width depending on the window size. But that's simply not feasible, because our version control systems (git) also depend on line breaks for automatic merging. I guess someone could write a git merge driver?

**Comment by j-f1** [comment]
Perhaps another reason we use 80-character-wide code is because narrower spans of text are more readable. For example, GitHub issue comments are about 80 characters wide, no matter how wide your monitor may be.

123456789**1**123456789**2**123456789**3**123456789**4**123456789**5**123456789**6**123456789**7**123456789**8**this wraps

(bolded numbers are tens)

The New York Times uses about 66:

<img width="300"  src="https://user-images.githubusercontent.com/25517624/43647452-acebb83c-9705-11e8-93a4-e403136d7086.png">

**Comment by Orodan** [comment]
Hi,

juste to share my opinion, I agree the way printWith is working now is the only reason why I don't use prettier in all my projects yet. 
For me, the core issue is actually not with how prettier is putting every arguments on a line that exceeds the width limit on their own line. It is with the fact that it forces the code to fill the width limit. Without this behavior, it would be much more easy to play with the witdth limit and to adapt it. I know it has already been mentionned and it would not be easy to implement in prettier, but that would be something awesome !

Anyway, love the prettier concept ! Keep up the good work :)

**Comment by j-f1** [comment]
We use the term “printing” to mean “formatting” @QuantumInformation.

**Comment by RomainLanz** [comment]
Hey all 👋 

I agree with @jogelin, @atombender.

Prettier has been created to avoid spending time thinking about how we should format code. But it nearly creates unreadable or very ugly code sometimes only because of this rule.

[This PR](https://github.com/prettier/prettier/pull/4765) is similar to this issue.

Prettier start to wrap your line even if you are inside the printWidth you gave him.

All the example given above shows this issue that keeps me and my team from using Prettier.

We are happy with nearly all the rule but not this behaviour that I believe should change.

Another example could be the following code:

```js
class TokensSchema extends Schema {
  up() {
    this.create('tokens', table => {
      table.increments()
      table.integer('user_id').unsigned().references('id').inTable('users')
      table.string('token', 255).notNullable().unique().index()
      table.string('type', 80).notNullable()
      table.boolean('is_revoked').defaultTo(false)
      table.timestamps()
    })
  }
}
```

It's readable and the longest line is 76 characters. Using Prettier it's transformed has:

```js
class TokensSchema extends Schema {
  up() {
    this.create('tokens', table => {
      table.increments()
      table
        .integer('user_id')
        .unsigned()
        .references('id')
        .inTable('users')
      table
        .string('token', 255)
        .notNullable()
        .unique()
        .index()
      table.string('type', 80).notNullable()
      table.boolean('is_revoked').defaultTo(false)
      table.timestamps()
    })
  }
}
```

Why that?

**Comment by RomainLanz** [comment]
I actually linked this PR on my message. 😄 

I was more thinking of having a way to disable this rule. I don't see any benefits (even the `git diff` argument is wrong for me) of obligating my team of having a line length. If they want to have the text to split they use the **word wrap** feature of their IDE/Text Editor.

**Comment by j-f1** [comment]
That’s one of the major features of Prettier @RomainLanz — it’s better than your editor’s soft-wrap functionality because it understands semantics and not just plain syntax.

**Comment by RomainLanz** [comment]
It understands semantics but it doesn't understand the context, which is by far much more important in my opinion.

I guess I'll see how it goes with the maximum `printWidth` when your PR is merged since it seems that change like that aren't welcome.

**Comment by lydell** [comment]
@RomainLanz `printWidth` is at the heart of Prettier. If you don't like that, Prettier might not be the tool you are looking for.

**Comment by slikts** [comment]
Quite a lot of (justified) griping about this issue here: https://github.com/prettier/prettier-vscode/issues/352

**Comment by asyncb** [comment]
Prettier doesn't mess with intentional vertical spacing, but it does with horizontal spacing.

1. I believe line length is up to the developer for deciding what actually looks good. If you can't deal with 80+ I suggest you learn how to use a computer or go back to 1980
2. printWidth shouldn't expand multi-lines into a long one, (if the printWidth is 120 for example)

**Comment by j-f1** [comment]
> I believe line length is up to the developer for deciding what actually looks good. If you can't deal with 80+ I suggest you learn how to use a computer or go back to 1980

@asyncb Note that even GitHub has an approximately 80-character-wide width for comments. This is because humans don’t do as well when reading extremely long lines. 

12345678911234567892123456789312345678941234567895123456789612345678971234567898123
(83 characters)

**Comment by j-f1** [comment]
> I believe line length is up to the developer for deciding what actually looks good. If you can't deal with 80+ I suggest you learn how to use a computer or go back to 1980

@asyncb Note that even GitHub has an approximately 80-character-wide width for comments. This is because humans don’t do as well when reading extremely long lines. 

12345678911234567892123456789312345678941234567895123456789612345678971234567898123
(83 characters)

**Comment by jdhines** [comment]
> you don’t write code formatted and then apply prettier, instead you write code in one line with no respect for formatting, press save and then prettier formats it.

I personally love Prettier for the other things it does, like quote style, tabs, indents, and semicolons, but I'm always fighting it on printWidth and wrapping, because like others have said, a set line length doesn't automatically make for readable code. There are times (like just now, having a config file with an array with 2 items, where it is more readable to have each on a line (since they're separate items), yet Prettier reformats it to a single line since it fits the limit, and then there's other times (like with an argument list) where I want them all to stay on one line because they represent similar things, and sometimes Prettier will split that out in ways that just take up more space.

So I think there needs to be a middle ground (like by being able to shut off wrapping) for people who want to take advantage of the all the other code style benefits of Prettier, but need to be able to make the call on when lines should be split or joined. Sure, it would mean that if I have wrapping turned off, someone could write an entire HTML template in one line, but that abuse would be a rare condition that could be handled individually. On the other hand, if I was on a project with hundreds of other devs, then sure, it might make sense to turn on wrapping.

**Comment by QuantumInformation** [comment]
Its more to help with git diffing and merging I feel.

**Comment by MrNovado** [comment]
While I do respect prettier for what it does, in principle, I think this particular rule is archaic and it needs to go. Especially in the modern-day and age when types and hooks/inline-selectors are a thing. Personally, I am more and more finding myself trying to escape prettier' formatting by using `// prettier-ignore`, but it is another sort of ugly and is not applicable in all cases.

```typescript
// original
const configuration = useSelector<isStateR, Configuration.AsObject>(s => s.configuration);
const initialLocalState = getInitialLocalState(configuration);
const [localState, localDispatch] = useReducer(getReducer(initialLocalState), initialLocalState);
const profileStateListeners = useRef<{ [index: string]: StateListener }>({});

// with prettier
const configuration = useSelector<isStateR, Configuration.AsObject>(
  s => s.configuration
);
const initialLocalState = getInitialLocalState(configuration);
const [localState, localDispatch] = useReducer(
  getReducer(initialLocalState),
  initialLocalState
);
const profileStateListeners = useRef<{ [index: string]: StateListener }>({});
```

```typescript
// original
const ProfileOptions = (
  props: DataVariants & 
  { initialState: LocalState; } &
  { dispatch: React.Dispatch<LocalActions> }
) => {

// with prettier
const ProfileOptions = (
  props: DataVariants & { initialState: LocalState } & {
    dispatch: React.Dispatch<LocalActions>;
  }
) => {
```

The hooks example is especially annoying because hooks allow you very neatly arrange your state and behaviors in one place (and you read them like columns: top-down, left-to-right), just for prettier to jumble it all over... Samish thing with memoized selectors (reselect' createSelector' body).

**TL;DR:** I would much prefer for prettier to respect my way of folding code. It can do anything else, but it should not try to break or unbreak lines.

Any escape hatch for this would be appreciated.
Also, `// prettier-ignore-start/end` does not work for me for whatever reason.

**Comment by slikts** [comment]
@MrNovado All you need to do is to set the `printWidth` config option to 100 (which is a fairly widespread preference; 120 is fine as well) and your Prettier-formatted example will look the same as the original. Secondarily, you're missing the whole point of Prettier.

**Comment by MrNovado** [comment]
@slikts 
> All you need to do is to set the `printWidth` config option to 100

This will solve absolutely nothing. While it will maybe work and not break lines in some places, it will unbreak lines in other places, like in jsx for example, making the change pointless or even more harmful. And what if I would require `printWidth` to be more or less than 100 at some point?

> Secondarily, you're missing the whole point of Prettier.

How very condescending. So enlighten me? Do you even have anything to add beyond what @vjeux already said here? The point of prettier is to automate formatting problems so you would not worry about them. And my point is, that whatever prettier does is not enough since I use it and it still makes me worry.

**Comment by slikts** [comment]
What's actually archaic is the 80 character print width, and your hooks example demonstrates that. The print width heuristic is still imperfect even with 100 and 120 characters, because a heuristic is a guess so it's wrong in some cases.

The missed point is automation; if automatic line formatting is not wanted, then `eslint --fix` can already be made to work the same as Prettier, except without automatic line formatting.

**Comment by gotenxds** [comment]
To "solve" this, we just work with a pre-hook commit for git that uses the default prettier options, as some us find 80 chars per line eligible and hard to read we just run prettier with different options on each pull and then again with the default on each commit. 
Honestly, I can't understand who can anyone read 3 break lines for a function call with 1 parameter, but to each his own 🤷‍♂

**Comment by dinofx** [comment]
Having the print width be relative to the current indentation level solves a lot of problems.  If 80 is the best heuristic, then shouldn't it be used regardless of where a block of code is nested/structured?

Having an absolute printWidth is not compatible with GitHub.  If I wrap existing code in a try/catch, unnecessarily modifying the code inside the block breaks line blame history.  It also increases the diff for reviewers.

Subconsciously, I'm sure some developers are "optimizing" their code to avoid indentation at the expense of something else.

**Comment by hoobss** [comment]
Honestly, I can't understand who can anyone read 3 break lines for a function call with 1 parameter,

**Comment by j-f1** [comment]
> If I wrap existing code in a try/catch, unnecessarily modifying the code inside the block breaks line blame history.

Won’t that happen anyway since you’re changing the indentation?

**Comment by dinofx** [comment]
> > If I wrap existing code in a try/catch, unnecessarily modifying the code inside the block breaks line blame history.
> 
> Won’t that happen anyway since you’re changing the indentation?

No, you set leading white space to be ignored when viewing diffs.  Gitlens in VSCode will do the same.

**Comment by thorn0** [comment]
@dinofx 

> Having the print width be relative to the current indentation level solves a lot of problems

There is a separate issue about this: https://github.com/prettier/prettier/issues/7064

**Comment by ackvf** [comment]
To be honest, having to set a small `printWidth`, because of a poorly implemented rule, and making it look as a good convention, instead of admitting that this well documented behavior is actually a bug is **shameful**.


> https://prettier.io/docs/en/options.html#print-width
>
>**For readability we recommend against using more than 80 characters:**
>
>In code styleguides, maximum line length rules are often set to 100 or 120. However, when humans write code, they don't strive to reach the maximum number of columns on every line. Developers often use whitespace to break up long lines for readability. In practice, the average line length often ends up well below the maximum.
>
>Prettier, on the other hand, strives to fit the most code into every line. With the print width set to 120, prettier may produce overly compact, or otherwise undesirable code.
>
>See the print width rationale for more information.


Please, focus especially on these sentence: 
>### Developers often use whitespace to break up long lines for readability.
>### Prettier, on the other hand, strives to fit the most code into every line.

**THIS DOESN'T EVEN MAKE SENSE**

A developer wouldn't ever do that unless they had a very good reason for it, so why does Prettier deliberately diverges from what is best for humans and intentionally formats the code to the opposite?

**Comment by lydell** [comment]
What the docs is trying to say is:

“If your style guide allows 120 columns, people won’t make _every_ line 120 columns. They’ll break lines earlier for readability.

Prettier’s printWidth option does not work the same way. It is _not_ the hard upper allowed line length limit. It is a way to say to Prettier roughly how long you’d like lines to be. Prettier will make both shorter and longer lines, but generally strive to meet the specified printWidth.

Remember, computers are dumb. You need to explicitly tell them what to do, while humans can make their own (implicit) judgements, for example on when to break a line.”

**Comment by aws-nlaquet** [comment]
> No, you set leading white space to be ignored when viewing diffs. Gitlens in VSCode will do the same.

But if you introduce a new level of indentation, say by wrapping your code in a try/catch, don't you need to re-evaluate your indentation for the existing code anyway, to make it conform to your style guidelines?

Prettier doesn't know the difference between you writing code for the first time, or adding in a try/catch block to existing code. It formats based off of what is currently present. If you don't re-apply your formatting to the newly indented code, then you're introducing formatting inconsistencies, and your formatting rules are broken.

This may make diffs more difficult; but it comes down to the core argument... A) should prettier break up lines that are too long, B) should humans decide how their lines are broken up. If A, then you get the current behavior and some diffs are less readable. If B, you lose a lot of the benefit of prettier, since you've just increased the cognitive overhead of formatting from practically 0 to back to near where it was before starting to use a tool like prettier.

Option C would be to tweak the heuristics of how prettier decides to break up lines, but that doesn't address the try/catch/diff problem.

**Comment by brianjonespfj** [comment]
> @lydell @vjeux @kachkaev My point is not to say "The code should not be split", my point is "To keep consistency, the code should be split based on syntax, not on print width". I think you will meet a lot of issue like #4172

Maybe just be a responsible developer and then you won't have to have a tool hard wrapping your code is wacky places.

**Comment by dinofx** [comment]
> But if you introduce a new level of indentation, say by wrapping your code in a try/catch, don't you need to re-evaluate your indentation for the existing code anyway, to make it conform to your style guidelines?

No.  The style guidelines would be relative to the current indent level, so existing code just shifts over one indent.

**Comment by or9** [comment]
But if your columns are longer than 80 characters, how do you expect to fit a line on a punch card? You'll never be able to deploy your code!  
http://pub.gajendra.net/2012/09/hollerith_tyranny  
And, apparently, [punch cards are based on the size of currency in the 1800's, so that they could use the same shipping envelopes for the 1890's census](https://softwareengineering.stackexchange.com/a/148678). So, basically it's anti-American do use a width other than 80 characters. Terrorists probably use 160 column width.
